test/Constraints/members.swift

```diff
@@ -74,7 +74,7 @@ struct GZ<T> {
 
 var z = Z(i: 0)
 var getI = z.getI
-var incI = z.incI // expected-error{{partial application of 'mutating'}}
+var incI = z.incI // expected-error{{calling a 'mutating' method without an argument list is not allowed}}
```

Comment:
I think using `calling` term here is not really correct because `z.incI` is not exactly call the method, a call to method would be `z.incI()`.  So maybe we can came up with another term for this like `accessing a 'mutating' method without an argument list is not allowed` or `using a 'mutating' ...` a note `did you mean to call it?` fix-it  "insert ()" if method don't require arguments or all have default values the same way we do in `downcast_to_unrelated_fixit`. @varungandhi-apple WDYT?

Reply:
Good point, `using` could be good.

Reply:
Luciano's suggestions sound good to me. 👍🏼

Reply:
In regards to adding a fix-it, it seems like I have to create a new diag message in the `DiagnosticsSema.def` file and in the `CSDiagnostics.cpp` file check if the method doesn't require arguments or has all default values then use the new diag message?

Is there a way I can append it to the existing diagnostics messages?
How can i perform the check on the arguments and default values (its been a while since ive done cpp)

Reply:
A fix-it can be attached to any diagnostic. In this case, I don't think we need a new fix-it, you can rename the `downcast_to_unrelated_fixit` to `add_empty_parameter_list_fixit` and reuse that here (since conceptually the idea is the same). So, the change would look something like:

```
  auto diagnostic = CompatibilityWarning
                        ? diag::partial_application_of_function_invalid_swift4
                        : diag::partial_application_of_function_invalid;

  emitDiagnosticAt(anchor->getNameLoc(), diagnostic, kind);
+ if (... checks pass ... ) {
+   emitDiagnosticAt(diagLoc, add_empty_parameter_list_fixit, funcName)
+    .fixitInsert(fixitLoc, "()");
+ }
```

You can take a look at how `downcast_to_unrelated_fixit` is emitted today to get the values for `diagLoc`, `fixitLoc` and `funcName`.

Please add a test case or two for the fix-it as well.

Reply:
Adding a fix-it would definitely be very useful! Also please note that to test newly added fix-its you could use {{<position>-<position>=<change>}}, there are a lot of examples of that in `test/Constraints` and `test/Sema`.

@LucianoPAlmeida  In regards to the diagnostic maybe it would be better to point out what could be done with a reference instead of what can't be e.g. `'mutating' method '<name>' can only be called` or `referencing 'mutating' method '<name>' is only possible via a call` or similar since `argument list` is kind of jargon-y especially considering that having argument list doesn't mean non-partial application :)

Reply:
> since argument list is kind of jargon-y especially considering that having argument list doesn't mean non-partial application

ah that is true because 
```
struct A {
  func f(p: Int) {}
}
let a = A()
let f = a.f(p:) // this could be also viewed as an "argument list"
```

> @LucianoPAlmeida In regards to the diagnostic maybe it would be better to point out what could be done with a reference instead of what can't be e.g. `'mutating' method '<name>' can only be called` or `referencing 'mutating' method '<name>' is only possible via a call`

Awesome, this looks definitely better! I like the later because `referencing` is looks a really good term for this :))

Reply:
Another option is "referencing 'mutating' method "foo" as a function value is not allowed"

Reply:
> Another option is "referencing 'mutating' method "foo" as a function value is not allowed"

I like this option =]

Reply:
Do we have any other `is not allowed` diagnostics? I think it’s good to rephrase “partial application” but something more actionable might be beneficial, if we can find it :)

Reply:
Thanks for the notes everyone, will get something done towards the end of the week

Reply:
Sounds great, thank you!

Reply:
> Do we have any other `is not allowed` diagnostics?

Another option could be `cannot reference  'mutating' method "foo" as a function value`, I think the actionable part can come from the note `did you mean to call it?` and then the fix-it to insert "()" if method don't require arguments or all have default values. WDYT?

Reply:
Yeah, that sounds good to me!